Add guardrails for remote model (#6750)
opensearch-project/documentation-website: _ml-commons-plugin/remote-models/guardrails.md

---
layout: default
title: Guardrails
has_children: false
has_toc: false
nav_order: 70
parent: Connecting to externally hosted models 
grand_parent: Integrating ML models
---

# Configuring model guardrails
**Introduced 2.13**
{: .label .label-purple }

> Guardrails can guide a large language model (LLM) toward desired behavior. They act as a filter, preventing the LLM from generating outputs that are harmful or violate ethical principles and facilitating a safer use of AI. Guardrails also steer the LLM to produce a more focused and relevant output. 

Reviewer: ```suggestion
Guardrails can guide a large language model (LLM) toward desired behavior. They act as a filter, preventing the LLM from generating output that is harmful or violates ethical principles and facilitating safer use of AI. Guardrails also cause the LLM to produce more focused and relevant output. 
```

To configure guardrails for your LLM, you can provide a list of words that are prohibited in the input or output of the model. Alternatively, you can provide a regular expression against which the model input or output will be matched.

## Prerequisites

Before you start, make sure you have fulfilled the [prerequisites]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/#prerequisites) for connecting to an externally hosted model.

## Step 1: Create a guardrail index

> To start, create an index that will hold the excluded words (_stopwords_). In the index settings, specify for the index to contain a `title` field, which will contain excluded words, and a `query` field, which is of the [percolator]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/percolator/) type. The percolator query will be used to match the LLM input or output:

Reviewer: ```suggestion
To start, create an index that will store the excluded words (_stopwords_). In the index settings, specify a `title` field, which will contain excluded words, and a `query` field, which is of the [percolator]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/percolator/) type. The percolator query will be used to match the LLM input or output:
```

## Step 2: Index excluded words or phrases

Next, index a query string query that will be used to match excluded words in the model input or output:

```json
PUT /words0/_doc/1?refresh
{
  "query": {
    "query_string": {
      "query": "title: blacklist"
    }
  }
}
```
{% include copy-curl.html %}

```json
PUT /words0/_doc/2?refresh
{
  "query": {
    "query_string": {
      "query": "title: \"Master slave architecture\""
    }
  }
}
```
{% include copy-curl.html %}

For more query string options, see [Query string query]({{site.url}}{{site.baseurl}}/query-dsl/full-text/query-string/).

## Step 3: Register a model group

To register a model group, send the following request:

```json
POST /_plugins/_ml/model_groups/_register
{
    "name": "bedrock",
    "description": "This is a public model group."
}
```
{% include copy-curl.html %}

The response contains the model group ID that you'll use to register a model to this model group:

```json
{
 "model_group_id": "wlcnb4kBJ1eYAeTMHlV6",
 "status": "CREATED"
}
```

To learn more about model groups, see [Model access control]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-access-control/).

## Step 4: Create a connector

> Create a connector to the model. In this example, you'll create a connector to the Anthropic Claude model hosted on Amazon Bedrock:

Reviewer 2: ```suggestion
Now you can create a connector for the model. In this example, you'll create a connector to the Anthropic Claude model hosted on Amazon Bedrock:
```

> Create a connector to the model. In this example, you'll create a connector to the Anthropic Claude model hosted on Amazon Bedrock:

Reviewer: "Now you can create a connector for the model"?
